### PR TITLE
core: Implement constant folding for AffineExpr

### PR DESCRIPTION
**Pull request by Groverkss**

### CODE REVIEW

File: tests/filecheck/parser-printer/affine_map.mlir
Comment (webmiche) [inline tests/filecheck/parser-printer/affine_map.mlir:35]: This looks wrong to me. Shouldn't the initial thing translate to 2+i? It seems to translate to 2-i, right?

File: xdsl/ir/affine/affine_expr.py
Comment (math-fehr) [inline xdsl/ir/affine/affine_expr.py:196]: That's so cursed ;) (But that's fine)

**Review by math-fehr — APPROVED** [review]
Nice! Thanks!

File: tests/filecheck/parser-printer/affine_map.mlir
Comment (superlopuh) [reply tests/filecheck/parser-printer/affine_map.mlir:35]: the whole thing is weird the expression in the check seems more complex than the one in the source...

**Review by superlopuh — COMMENTED** [review]
I don't really understand, it seems like the output is more complex than the source. Should we not aim to at least make it as simple as the input, and maybe simpler?

File: tests/filecheck/parser-printer/affine_map.mlir
Comment (Groverkss) [reply tests/filecheck/parser-printer/affine_map.mlir:35]: Oops, this was a parser bug. Fixed now.

File: xdsl/ir/affine/affine_expr.py
Comment (superlopuh) [inline xdsl/ir/affine/affine_expr.py:162]: Before you merge this, why not make the signature of this function 
```suggestion
    def _try_fold_constant(self) -> AffineExpr:
```
Where you return self if you could not merge the constant? Where the implementation folds the constant in the lhs, rhs, and then matches on the kind of the operation to see how they could be combined?

Comment (superlopuh) [reply xdsl/ir/affine/affine_expr.py:162]: Also it seems like swapping around the order of arguments might be confusing, feels like that's a separate thing from constant folding to me.

**Review by superlopuh — CHANGES_REQUESTED** [review]

File: xdsl/ir/affine/affine_expr.py
Comment (Groverkss) [reply xdsl/ir/affine/affine_expr.py:208]: Any better way to swap things?

**Comment by Groverkss** [comment]
> I don't really understand, it seems like the output is more complex than the source. Should we not aim to at least make it as simple as the input, and maybe simpler?

This will be done as more simplifications are added. This patch just adds some of them.

File: xdsl/ir/affine/affine_expr.py
Comment (Groverkss) [reply xdsl/ir/affine/affine_expr.py:162]: > Also it seems like swapping around the order of arguments might be confusing, feels like that's a separate thing from constant folding to me.

It's canonicalizing the input for simplifications. Seems pretty standard to me?

Comment (Groverkss) [reply xdsl/ir/affine/affine_expr.py:162]: > Before you merge this, why not make the signature of this function
> 
> Where you return self if you could not merge the constant? Where the implementation folds the constant in the lhs, rhs, and then matches on the kind of the operation to see how they could be combined?

I like the current implementation. It allows me to fold things like (expr + constant) + constant too.

**Comment by Groverkss** [comment]
I added some more simplifications for now, but I don't think I'll add any more simplifications for now.

**Review by superlopuh — APPROVED** [review]

File: xdsl/ir/affine/affine_expr.py
Comment (math-fehr) [reply xdsl/ir/affine/affine_expr.py:208]: No that's the best!
The only curse thing in my opinion is swapping `self`, but that's fine!

Comment (webmiche) [reply xdsl/ir/affine/affine_expr.py:208]: Maybe something like `other.__add__(self)`?

**Pull request merged**
